test/forge/CarbonPOL.t.sol

```diff
@@ -0,0 +1,744 @@
+        i = bound(i, 0, 2);
+        // enable trading
+        vm.prank(admin);
+        carbonPOL.enableTrading(tokens[i], ICarbonPOL.Price({ ethAmount: 100, tokenAmount: 10000 }));
+        // assert trading is enabled
+        assertTrue(carbonPOL.tradingEnabled(tokens[i]));
+    }
+
+    /// @dev test enabling trading for a token should emit an event
+    function testEnablingTradingForTokenShouldEmitAnEvent(uint256 i) public {
+        // pick one of these tokens to test
+        Token[3] memory tokens = [token1, token2, bnt];
+        // pick a random number from 0 to 2 for the tokens
+        i = bound(i, 0, 2);
+        ICarbonPOL.Price memory price = ICarbonPOL.Price({ ethAmount: 100, tokenAmount: 10000 });
+        vm.prank(admin);
+        // expect event to be emitted
+        vm.expectEmit();
+        emit TradingEnabled(tokens[i], price);
+        // enable trading
+        carbonPOL.enableTrading(tokens[i], price);
+    }
+
+    /// @dev test should revert when setting invalid price for a token
+    function testShouldRevertWhenSettingInvalidPriceForToken(uint256 i) public {
+        // pick one of these tokens to test
+        Token[3] memory tokens = [token1, token2, bnt];
+        // pick a random number from 0 to 2 for the tokens
+        i = bound(i, 0, 2);
+        // enable trading
+        vm.startPrank(admin);
+        // setting any of ethAmount or tokenAmount to 0 results in invalid price error
+        vm.expectRevert(ICarbonPOL.InvalidPrice.selector);
+        carbonPOL.enableTrading(tokens[i], ICarbonPOL.Price({ ethAmount: 0, tokenAmount: 10000 }));
+        vm.expectRevert(ICarbonPOL.InvalidPrice.selector);
+        carbonPOL.enableTrading(tokens[i], ICarbonPOL.Price({ ethAmount: 100000, tokenAmount: 0 }));
+        vm.expectRevert(ICarbonPOL.InvalidPrice.selector);
+        carbonPOL.enableTrading(tokens[i], ICarbonPOL.Price({ ethAmount: 0, tokenAmount: 0 }));
+    }
+
+    /// @dev test should properly return price for enabled tokens as time passes
+    function testShouldProperlyReturnPriceAsTimePasses(uint128 ethAmount, uint128 tokenAmount, uint256 i) public {
+        // pick one of these tokens to test
+        Token[3] memory tokens = [token1, token2, bnt];
+        // pick a random number from 0 to 2 for the tokens
+        i = bound(i, 0, 2);
+        // enable trading and set price for token
+        vm.prank(admin);
+        Token token = tokens[i];
+
+        ethAmount = uint128(bound(ethAmount, 10, 1e30));
+        tokenAmount = uint128(bound(tokenAmount, 10, 1e30));
+
+        ICarbonPOL.Price memory initialPrice = ICarbonPOL.Price({ ethAmount: ethAmount, tokenAmount: tokenAmount });
+        carbonPOL.enableTrading(token, initialPrice);
+
+        // set timestamp to 1
+        vm.warp(1);
+
+        ICarbonPOL.Price memory price = carbonPOL.tokenPrice(token);
+        uint128 expectedETHAmount = getExpectedETHAmount(ethAmount);
+
+        assertEq(price.ethAmount, expectedETHAmount);
```

Comment:
this tests the contract vs itself which basically means that it doesn't really test anything - I recommend creating some hardcoded scenarios and testing the values against expected (hardcoded) results in those scenarios

Reply:
added test cases for different initial prices and timestamps